[ty] introduce local variables for `from` imports of submodules in `__init__.py(i)`

Gankra opened the pull request:
This rips out the previous implementation in favour of a new implementation with 3 rules:

- **froms are locals**: a `from..import` can only define locals, it does not have global
    side-effects. Specifically any submodule attribute `a` that's implicitly introduced by either
    `from .a import b` or `from . import a as b` (in an `__init__.py(i)`) is a local and not a
    global. If you do such an import at the top of a file you won't notice this. However if you do
    such an import in a function, that means it will only be function-scoped (so you'll need to do
    it in every function that wants to access it, making your code less sensitive to execution
    order).

- **first from first serve**: only the *first* `from..import` in an `__init__.py(i)` that imports a
    particular direct submodule of the current package introduces that submodule as a local.
    Subsequent imports of the submodule will not introduce that local. This reflects the fact that
    in actual python only the first import of a submodule (in the entire execution of the program)
    introduces it as an attribute of the package. By "first" we mean "the first time in this scope
    (or any parent scope)". This pairs well with the fact that we are specifically introducing a
    local (as long as you don't accidentally shadow or overwrite the local).

- **dot re-exports**: `from . import a` in an `__init__.pyi` is considered a re-export of `a`
    (equivalent to `from . import a as a`). This is required to properly handle many stubs in the
    wild. Currently it must be *exactly* `from . import ...`.
    
This implementation is intentionally limited/conservative (notably, often requiring a from import to be relative). I'm going to file a ton of followups for improvements so that their impact can be evaluated separately.


Fixes https://github.com/astral-sh/ty/issues/133

Gankra commented:
Preserving for future comparison: this is the "submodule-attributes" approach's results. I am now pushing up a different WIP implementation.

<img width="1053" height="800" alt="Screenshot 2025-11-03 at 2 41 32 PM" src="https://github.com/user-attachments/assets/a63e783f-b3a7-4689-9011-2d89a7b6bf27" />

sharkdp commented:
> Preserving for future comparison

Just FYI, you can also access the full history here, I do that quite often after pushing a change.

<img width="1102" height="594" alt="image" src="https://github.com/user-attachments/assets/9dc84d7f-b127-4113-a853-cda4e9f19527" />

Gankra commented:
I have cleaned up the implementation and added some more tests. I've made a few edge-cases a bit more strict. I am hoping this is a complete implementation, except for some code/mdtest comments need to be updated to reflect the current semantics.

For posterity, this is the results that we agreed seemed shippable *before* my cleanup. Hopefully the stricter implementation isn't a huge change!

<img width="937" height="608" alt="Screenshot 2025-11-04 at 3 41 00 PM" src="https://github.com/user-attachments/assets/b05d03b1-a0a2-4576-bc1c-986be26aa4fb" />

carljm commented on crates/ty_python_semantic/resources/mdtest/import/nonstandard_conventions.md:
```diff
@@ -1,39 +1,35 @@
+    global. If you do such an import at the top of a file you won't notice this. However if you do
+    such an import in a function, that means it will only be function-scoped (so you'll need to do
+    it in every function that wants to access it, making your code less sensitive to execution
+    order).
```
I assume that in the upcoming PR you mentioned in person, this whole paragraph goes away to be replaced by something that says "we only model submodule imports adding the submodule attribute for global-scope submodule imports"

carljm commented on crates/ty_python_semantic/resources/mdtest/import/nonstandard_conventions.md:
```diff
@@ -1,39 +1,35 @@
-The `from . import submodule` idiom in an `__init__.pyi` is fairly explicit and we should definitely
-support it.
+The `from . import submodule` idiom in an `__init__.pyi` should definitely be considered an explicit
+re-export.
```
Nit: "should definitely be" seems like quite strong language here, given that all of the stub-file rules are kind of arbitrary (there's no runtime behavior to model it on) and we're just trying to find the right balance between "what the spec explicitly says" and "what stub files in the wild rely on." I'd probably just give this a more factual tone: "We consider ... to be an explicit re-export."

carljm commented on crates/ty_python_semantic/resources/mdtest/import/nonstandard_conventions.md:
```diff
@@ -1,39 +1,35 @@
+    (equivalent to `from . import a as a`). This is required to properly handle many stubs in the
+    wild. Currently it must be *exactly* `from . import ...`.
 
 ## Relative `from` Import of Direct Submodule in `__init__`
```
Nit: the fact that all test headers / names in this test file assume "in a stub file" by default (and then add "(Non-stub Check)" to the equivalent non-stub test) may have made sense before this PR, when this file was only about modifying stub file rules, but it feels weird after this PR. I would probably put the non-stub test first in each scenario, and explicitly mark the stub tests instead? (Or really I think I would just have each test header once and then use sub-headings `### In py file` and `### In stub file` rather than repeating the whole test description.

Gankra replied:
You're extremely right and I was hoping no one would notice. *rolls up sleeves*

carljm commented on crates/ty_python_semantic/resources/mdtest/import/nonstandard_conventions.md:
```diff
@@ -752,9 +764,8 @@ Can easily result in the typechecker getting "confused" and thinking imports of
-more powerful than that in the current implementation strategy quickly gets the functions and
-submodules mixed up.
+We avoid this by ~desugarring the imports into two definitions, with the RHS occuring second, so
+that it can freely overwrite the LHS.
```
```suggestion
We avoid this by ensuring that the imported name (the right-hand `funcmod` in `from .funcmod import funcmod`) overwrites the submodule attribute (the left-hand `funcmod`), as it does at runtime.
```

carljm commented on crates/ty_python_semantic/resources/mdtest/type_properties/is_equivalent_to.md:
````diff
@@ -607,23 +607,33 @@ module:
+import imported.abc
+```
+
+`imported/__init__.pyi`:
````
I think it's fine (even good) to make this change, but I assume the previous version of this test (using `importlib`) still passes after this PR? It's just that the simpler version was easier to debug in development?

Gankra replied:
Different versions of this PR have variously broken and unbroken the test because of the line `from importlib.abc import Loader` in typeshed's `importlib/__init__.pyi`. I would rather check in this not-relying-on-a-random-library version.

carljm commented on crates/ty_python_semantic/src/semantic_index/builder.rs:
```diff
@@ -1453,6 +1449,43 @@ impl<'ast> Visitor<'ast> for SemanticIndexBuilder<'_, 'ast> {
+                    let symbol = self.add_symbol(direct_submodule_name);
+                    self.add_definition(
+                        symbol.into(),
+                        ImportFromImplicitDefinitionNodeRef { node, submodule },
```
I don't know that it matters, since the from-import node itself is otherwise unused as a definition NodeKey, and there can only be one implicit definition per from-import, but I'm curious whether we could attach this definition to the Identifier for the submodule itself rather than to the entire import-from node? It seems slightly more intuitive/clearer. Maybe we can't do this because `Identifier` isn't a real AST node? (I don't recall...)

carljm commented on crates/ty_python_semantic/src/types/infer/builder.rs:
```diff
@@ -5475,6 +5483,99 @@ impl<'db, 'ast> TypeInferenceBuilder<'db, 'ast> {
+        // will ever set `x`, and any subsequent overwrites of it will permanently clobber it.
+        // Also, a local variable `x` in a function should always shadow the submodule because
+        // the submodule is defined at file-scope. However, both of these issues are much more
+        // narrow, so this approach seems to work well in practice!)
```
As discussed in person, I don't think accepting these two bugs is the right tradeoff, vs just not modeling submodule-attribute side-effects at all for imports inside functions. Ok with landing this as-is, as long as there is an upcoming subsequent PR that adjusts this to simply not create these "implicit" definitions at all inside functions.

carljm commented on crates/ty_python_semantic/src/types/infer/builder.rs:
```diff
@@ -5475,6 +5483,99 @@ impl<'db, 'ast> TypeInferenceBuilder<'db, 'ast> {
+        };
+
+        let diagnostic = builder.into_diagnostic(format_args!(
+            "Module `{module_name}` has no member `{name}`"
```
Nit: we are trying to create the "member" here -- what we've failed to do is find the submodule that it should point to. So I would expect something more like:
```suggestion
            "Module `{module_name}` has no submodule `{name}`"
```

Gankra commented on crates/ty_python_semantic/src/types/infer/builder.rs:
```diff
@@ -5475,6 +5483,99 @@ impl<'db, 'ast> TypeInferenceBuilder<'db, 'ast> {
+
+        if !self.is_reachable(import_from) {
+            return;
+        }
```
No this is copy-paste from `infer_from_import_definition`, I have no insights. I just assumed it was a common "don't do diagnostics about dead code" idiom.

Gankra replied:
Followup it is!
